texar/modules/pretrained/xlnet_model_utils.py

```diff
@@ -0,0 +1,342 @@
+        output = self.dropout(output)
+        # residual + layer norm
+        output = self.layer_norm(input + output)
+        return output
+
+
+class PositionalEmbedding(nn.Module):
+
+    def __init__(self, embed_dim: int):
+        super().__init__()
+
+        freq_seq = torch.arange(0.0, embed_dim, 2.0)
+        inv_freq = 1 / (10000 ** (freq_seq / embed_dim))
+        self.register_buffer('inv_freq', inv_freq)
+
+    def forward(self,  # type: ignore
+                pos_seq: torch.Tensor) -> torch.Tensor:
+        sinusoid = torch.ger(pos_seq, self.inv_freq)
+        pos_embed = torch.cat([sinusoid.sin(), sinusoid.cos()], dim=-1)
+        return pos_embed
+
+
+class RelativePositionalEncoding(ModuleBase):
+
+    def __init__(self, hparams=None):
+        super().__init__(hparams)
+        self.sinusoid_embed = PositionalEmbedding(self._hparams.dim)
+
+    @staticmethod
+    def default_hparams():
+        return {
+            "dim": 768,
+            "max_seq_len": 512,
+        }
+
+    def _create_positional_embedding(self,
+                                     start: int,
+                                     end: int,
+                                     step: int,
+                                     batch_size: int,
+                                     clamp_len: Optional[int] = None) \
+            -> torch.Tensor:
+        embed_buffer = next(self.sinusoid_embed.buffers())
+        pos_seq = torch.arange(start, end, step, device=embed_buffer.device,
+                               dtype=embed_buffer.dtype)
+
+        if clamp_len is not None:
+            pos_seq = torch.clamp(pos_seq, -clamp_len, clamp_len)
+
+        pos_embed = self.sinusoid_embed(pos_seq)
+        pos_embed = pos_embed.unsqueeze(1).expand(-1, batch_size, -1)
+        return pos_embed
+
+    def forward(self,  # type: ignore
+                batch_size: int,
+                seq_len: int,
+                total_len: int,
+                clamp_len: Optional[int] = None,
+                attn_type: str = 'bi',
+                bi_data: bool = True) -> torch.Tensor:
+        if attn_type == 'bi':
+            start, end = total_len, -seq_len
+        elif attn_type == 'uni':
+            start, end = total_len, -1
+        else:
+            raise ValueError(f"Unknown `attn_type` {attn_type}")
+
+        if bi_data:
+            if batch_size % 2 != 0:
+                raise ValueError("`batch_size` must be an even number")
+            fwd_pos_embed = self._create_positional_embedding(
+                start, end, -1, batch_size // 2, clamp_len)
+            bwd_pos_embed = self._create_positional_embedding(
+                -start, -end, 1, batch_size // 2, clamp_len)
+            pos_embed = torch.cat([fwd_pos_embed, bwd_pos_embed], dim=1)
+        else:
+            pos_embed = self._create_positional_embedding(
+                start, end, -1, batch_size, clamp_len)
+        return pos_embed
+
+
+class RelativeMultiheadAttention(ModuleBase):
```

Comment:
Is this module possibly applicable in other models? What's the difference from `MultiheadAttentionEncoder`? We can discuss this in person.

It's one of the key advantages of Texar over other libraries that we correctly factorize out any re-usable components and make them as standardized modules (e.g., the above new embedders)

Reply:
This is the relative attention module as in TransformerXL, with adaptations for XLNet (segment embeddings and two-stream attention).

The difference between this and the vanilla multi-head attention is mainly two-fold:
1) it uses relative position embeddings (though in current implementation it's not computed in this module);
2) it attends not only to previous hidden states (`states_h`) but also to (detached) hidden states from previous batches (`memory`).

If necessary we could probably add a version without XLNet-specific adaptations to Texar modules.

Reply:
I think `Segment embedding` should not be included in this module. Like in BERT, all embeddings, including word embedding, segment embedding, and position embedding, are separated from `MultiheadAttentionEncoder`. They are aggregated externally as a single embedding before feeding to the encoder. Can we do the same factorization here?

For `two-stream attention`, we can probably have an option to toggle b/w single-stream or two-stream? Or at this point we can simply have a two-stream version for XLNet

Let's make the module a texar module

Reply:
The difference here is that positional embeddings and segment embeddings are directly fed into each layer, while word embeddings are only used as the input to the first layer. So this module must take pos & seg embeddings as arguments.

Now two-stream attention is disabled when `states_g` is `None`.

I just think that this is a very model-specific implementation and might not be useful elsewhere. I'm more in favor of adding a general relative attention module (as in TransformerXL) to Texar modules, and keep this as is.